scrypto-compiler/src/lib.rs

```diff
@@ -0,0 +1,910 @@
+            target_path.push("target");
+            target_path
+        } else if let Some(directory) = &input_params.target_directory {
+            // If target directory is explicitly specified as compiler parameter then use it as is
+            PathBuf::from(directory)
+        } else {
+            // If target directory is not specified as compiler parameter then get default
+            // target directory basing on manifest file
+            PathBuf::from(&Self::get_default_target_directory(&manifest_path)?)
+        };
+
+        let mut target_binary_directory = target_directory.clone();
+        target_binary_directory.push(BUILD_TARGET);
+        target_binary_directory.push(input_params.profile.as_directory_name());
+
+        let target_binary_path =
+            Self::get_target_binary_path(&manifest_path, &target_binary_directory)?;
+
+        Ok((manifest_path, target_directory, target_binary_path))
+    }
+
+    fn prepare_command(&mut self, command: &mut Command) -> Result<(), ScryptoCompilerError> {
+        let features_list = self.prepare_features();
+        let features = (!features_list.is_empty())
+            .then_some(vec!["--features", &features_list])
+            .unwrap_or_default();
+
+        let rustflags = self.prepare_rust_flags();
+
+        let package = self
+            .input_params
+            .package
+            .as_ref()
+            .and_then(|p| Some(vec!["--package", &p]))
+            .unwrap_or_default();
+
+        if self.input_params.coverage {
+            // coverage uses '-Z' flag which requires use of nightly toolchain
+            command.arg("+nightly");
+        }
+
+        command
+            .arg("build")
+            .arg("--target")
+            .arg(BUILD_TARGET)
+            .args(self.input_params.profile.as_command_args())
+            .arg("--target-dir")
+            .arg(&self.target_directory)
+            .arg("--manifest-path")
+            .arg(&self.manifest_path)
+            .args(package)
+            .args(features)
+            .env("CARGO_ENCODED_RUSTFLAGS", rustflags)
+            .envs(self.input_params.set_environment_variables.clone());
+
+        self.input_params
+            .unset_environment_variables
+            .iter()
+            .for_each(|e| {
+                command.env_remove(e);
+            });
+
+        Ok(())
+    }
+
+    fn wasm_optimize(&mut self) -> Result<(), ScryptoCompilerError> {
+        if let Some(wasm_opt_config) = &self.input_params.wasm_optimization {
+            wasm_opt_config
+                .run(&self.target_binary_path, &self.target_binary_path)
+                .map_err(ScryptoCompilerError::WasmOptimizationError)
+        } else {
+            Ok(())
+        }
+    }
+
+    // Returns output wasm file path
+    pub fn compile(&mut self) -> Result<PathBuf, ScryptoCompilerError> {
+        // Create compilation command
+        let mut command = Command::new("cargo");
```

Comment:
I've implemented this by additional compile function: `compile_with_stdio` because `Stdio` object is not implementing `Clone`, so I cannot store it in `InputParams` struct.